File: imagepicker/src/main/java/com/weimu/library/view/ImageSelectorActivity.java
package com.weimu.library.view;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.RequiresApi;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.weimu.library.ImageStaticHolder;
import com.weimu.library.R;
import com.weimu.library.adapter.ImageFolderAdapter;
import com.weimu.library.adapter.ImageListAdapter;
import com.weimu.library.core.StatusManager;
import com.weimu.library.core.ToolBarManager;
import com.weimu.library.model.LocalMedia;
import com.weimu.library.model.LocalMediaFolder;
import com.weimu.library.utils.FileUtilsIP;
import com.weimu.library.utils.GridSpacingItemDecoration;
import com.weimu.library.utils.LocalMediaLoader;
import com.weimu.library.utils.ScreenUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import top.zibin.luban.Luban;
import top.zibin.luban.OnCompressListener;


public class ImageSelectorActivity extends SelectorBaseActivity {
    public final static int REQUEST_IMAGE = 66;
    public final static int REQUEST_CAMERA = 67;

    public final static String BUNDLE_CAMERA_PATH = "CameraPath";

    public final static String REQUEST_OUTPUT = "outputList";


    public final static String EXTRA_MAX_SELECT_NUM = "MaxSelectNum";//最大选择数
    public final static String EXTRA_SELECT_MODE = "SelectMode";//选择模式
    public final static String EXTRA_SHOW_CAMERA = "ShowCamera";//是否显示摄像头
    public final static String EXTRA_ENABLE_PREVIEW = "EnablePreview";//是否需要预览
    public final static String EXTRA_ENABLE_CROP = "EnableCrop";//是否需要裁剪
    public final static String EXTRA_ENABLE_COMPRESS = "EnableCompress";//是否需要压缩

    public final static int MODE_MULTIPLE = 1;
    public final static int MODE_SINGLE = 2;

    private int maxSelectNum = 9;
    private int selectMode = MODE_MULTIPLE;
    private boolean enableCamera = true;
    private boolean enablePreview = true;
    private boolean enableCrop = false;
    private boolean enableCompress = false;//是否显示原图按钮

    private int spanCount = 4;
    //ui
    private RecyclerView recyclerView;
    private ImageListAdapter imageAdapter;
    private LinearLayout folderLayout;
    private TextView folderName;
    private FolderWindow folderWindow;
    private CheckBox cbOrigin;
    private ToolBarManager toolBarManager;

    private String cameraPath;

    private List<LocalMediaFolder> allFolders;//所有图片文件夹

    private boolean isUseOrigin = false;//是否使用原图

    public static void start(Activity activity, int maxSelectNum, int mode, boolean enableCamera, boolean enablePreview, boolean enableCrop, boolean enableCompress) {
        Intent intent = new Intent(activity, ImageSelectorActivity.class);
        intent.putExtra(EXTRA_MAX_SELECT_NUM, maxSelectNum);
        intent.putExtra(EXTRA_SELECT_MODE, mode);
        intent.putExtra(EXTRA_SHOW_CAMERA, enableCamera);
        intent.putExtra(EXTRA_ENABLE_PREVIEW, enablePreview);
        intent.putExtra(EXTRA_ENABLE_CROP, enableCrop);
        intent.putExtra(EXTRA_ENABLE_COMPRESS, enableCompress);
        activity.startActivityForResult(intent, REQUEST_IMAGE);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_imageselector);

        maxSelectNum = getIntent().getIntExtra(EXTRA_MAX_SELECT_NUM, 9);
        selectMode = getIntent().getIntExtra(EXTRA_SELECT_MODE, MODE_MULTIPLE);
        enableCamera = getIntent().getBooleanExtra(EXTRA_SHOW_CAMERA, true);
        enablePreview = getIntent().getBooleanExtra(EXTRA_ENABLE_PREVIEW, true);
        enableCrop = getIntent().getBooleanExtra(EXTRA_ENABLE_CROP, false);
        enableCompress = getIntent().getBooleanExtra(EXTRA_ENABLE_COMPRESS, false);


        if (selectMode == MODE_MULTIPLE) {
            enableCrop = false;
        } else {
            enablePreview = false;
        }
        if (savedInstanceState != null) {
            cameraPath = savedInstanceState.getString(BUNDLE_CAMERA_PATH);
        }
        initView();
        registerListener();

        ///load data
        new LocalMediaLoader(this, LocalMediaLoader.TYPE_IMAGE).loadAllImage(new LocalMediaLoader.LocalMediaLoadListener() {

            @Override
            public void loadComplete(List<LocalMediaFolder> folders) {
                allFolders = folders;
                folderWindow.bindFolder(allFolders);
                //load all images first
                imageAdapter.bindImages(allFolders.get(0).getImages());
            }
        });

    }

    public void initView() {
        StatusManager.getInstance().setColor(this, R.color.white);
        toolBarManager = ToolBarManager.with(this, getContentView())
                .setBackgroundColor(R.color.white)
                .setTitle("选择图片")
                .setNavigationIcon(R.drawable.toolbar_arrow_back_black)
                .setMenuTextContent(getString(R.string.done))
                .setMenuTextColors(R.color.black_text_selector)
                .setMenuTextEnable(false)
                .setMenuTextClick(new ToolBarManager.OnMenuTextClickListener() {
                    @Override
                    public void onMenuTextClick() {
                        //点击完成
                        onSelectDone(imageAdapter.getSelectedImages());
                    }
                });


        folderWindow = new FolderWindow(this);

        //todo 完成按钮
        if (selectMode == MODE_MULTIPLE) {
            toolBarManager.setMenuTextContent(getString(R.string.done));
        } else {
            toolBarManager.setMenuTextContent("");
        }


        //是否使用原图
        cbOrigin = findViewById(R.id.cb_origin);

        if (!enableCompress) {
            cbOrigin.setVisibility(View.GONE);
        }

        cbOrigin.setChecked(isUseOrigin);
        cbOrigin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                isUseOrigin = !cbOrigin.isChecked();
            }
        });

        folderLayout = (LinearLayout) findViewById(R.id.folder_layout);
        folderName = (TextView) findViewById(R.id.folder_name);

        recyclerView = (RecyclerView) findViewById(R.id.folder_list);
        recyclerView.setHasFixedSize(true);
        recyclerView.addItemDecoration(new GridSpacingItemDecoration(spanCount, ScreenUtils.dip2px(this, 2), false));
        recyclerView.setLayoutManager(new GridLayoutManager(this, spanCount));

        imageAdapter = new ImageListAdapter(this, maxSelectNum, selectMode, enableCamera, enablePreview);
        recyclerView.setAdapter(imageAdapter);
    }

    public void registerListener() {
        folderLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Toast.makeText(ImageSelectorActivity.this, "文件夹长度  " + allFolders.size() + "  内部图片数量  " + allFolders.get(0).getImages().size(), Toast.LENGTH_SHORT).show();
                if (allFolders.size() == 0 || allFolders.get(0).getImages().size() == 0) {
                    Toast.makeText(ImageSelectorActivity.this, "没有可选择的图片", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (folderWindow.isShowing()) {
                    folderWindow.dismiss();
                } else {
                    folderWindow.showAsDropDown(toolBarManager.getToolBarView());
                }
            }
        });
        //recyclerView点击事件
        imageAdapter.setOnImageSelectChangedListener(new ImageListAdapter.OnImageSelectChangedListener() {
            @Override
            public void onChange(List<LocalMedia> selectImages) {
                boolean enable = selectImages.size() != 0;
                toolBarManager.setMenuTextEnable(enable);
                if (enable) {
                    toolBarManager.setMenuTextContent(getString(R.string.done_num, selectImages.size() + "", maxSelectNum + ""));
                } else {
                    toolBarManager.setMenuTextContent(getString(R.string.done));
                }
            }

            @Override
            public void onTakePhoto() {
                startCamera();
            }

            @Override
            public void onPictureClick(LocalMedia media, int position, View view) {
                if (enablePreview) {

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        startPreviewWithAnim(imageAdapter.getImages(), position, view);
                    } else {
                        startPreview(imageAdapter.getImages(), position);
                    }

                } else if (enableCrop) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        startCropWithAnim(media.getPath(), view);
                    } else {
                        startCrop(media.getPath());
                    }
                } else {
                    onSelectDone(media.getPath());
                }
            }
        });
        //点击某个文件件
        folderWindow.setOnItemClickListener(new ImageFolderAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(String name, List<LocalMedia> images) {
                folderWindow.dismiss();
                imageAdapter.bindImages(images);
                folderName.setText(name);
                recyclerView.smoothScrollToPosition(0);
            }
        });
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            // on take photo success
            if (requestCode == REQUEST_CAMERA) {
                sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(new File(cameraPath))));
                if (enableCrop) {
                    startCrop(cameraPath);
                } else {
                    onSelectDone(cameraPath);
                }
            }
            //on preview select change
            else if (requestCode == ImagePreviewActivity.REQUEST_PREVIEW) {
                boolean isDone = data.getBooleanExtra(ImagePreviewActivity.OUTPUT_ISDONE, false);
                List<LocalMedia> images = (List<LocalMedia>) data.getSerializableExtra(ImagePreviewActivity.OUTPUT_LIST);
                if (isDone) {
                    onSelectDone(images);
                } else {
                    imageAdapter.bindSelectImages(images);
                }
            }
            // on crop success
            else if (requestCode == ImageCropActivity.REQUEST_CROP) {
                String path = data.getStringExtra(ImageCropActivity.OUTPUT_PATH);
                onSelectDone(path);
            }
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(BUNDLE_CAMERA_PATH, cameraPath);
    }

    /**
     * start to camera、preview、crop
     */
    public void startCamera() {
        File cameraFile = FileUtilsIP.createCameraFile(this);
        cameraPath = cameraFile.getAbsolutePath();
        FileUtilsIP.startActionCapture(this, cameraFile, REQUEST_CAMERA);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public void startPreviewWithAnim(List<LocalMedia> previewImages, int position, View view) {
        ImagePreviewActivity.startPreviewWithAnim(this, previewImages, imageAdapter.getSelectedImages(), maxSelectNum, position, view);
    }

    public void startPreview(List<LocalMedia> previewImages, int position) {
        ImagePreviewActivity.startPreview(this, previewImages, imageAdapter.getSelectedImages(), maxSelectNum, position);
    }

    @SuppressLint("RestrictedApi")
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public void startCropWithAnim(String path, View view) {
        startActivityForResult(ImageCropActivity.newIntent(this, path), ImageCropActivity.REQUEST_CROP,
                ActivityOptions.makeSceneTransitionAnimation(this, view, "share_image").toBundle());
    }

    public void startCrop(String path) {
        startActivityForResult(ImageCropActivity.newIntent(this, path), ImageCropActivity.REQUEST_CROP);
    }

    /**
     * on select done
     *
     * @param medias
     */
    public void onSelectDone(List<LocalMedia> medias) {
        ArrayList<String> images = new ArrayList<>();
        for (LocalMedia media : medias) {
            images.add(media.getPath());
        }
        onResult(images);
    }

    public void onSelectDone(String path) {
        ArrayList<String> images = new ArrayList<>();
        images.add(path);
        onResult(images);
    }

    //返回图片
    public void onResult(ArrayList<String> images) {
        if (isUseOrigin) {
            setResult(RESULT_OK, new Intent().putStringArrayListExtra(REQUEST_OUTPUT, images));
            finish();
        } else {
            compressImage(images);

        }
    }

    //压缩图片
    private void compressImage(final ArrayList<String> photos) {
        //Toast.makeText(this, "压缩中...", Toast.LENGTH_SHORT).show();
        final List<String> newImageList = new ArrayList<>();
        Luban.with(this)
                .load(photos)                                   // 传人要压缩的图片列表
                .ignoreBy(100)                                  // 忽略不压缩图片的大小
                .setCompressListener(new OnCompressListener() { //设置回调
                    @Override
                    public void onStart() {
                        Log.d("weimu", "开始压缩");
                    }

                    @Override
                    public void onSuccess(File file) {
                        Log.d("weimu", "压缩成功 地址为：" + file.toString());
                        newImageList.add(file.toString());
                        //所有图片压缩成功
                        if (newImageList.size() == photos.size()) {
                            setResult(RESULT_OK, new Intent().putStringArrayListExtra(REQUEST_OUTPUT, photos));
                            finish();
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }
                }).launch();    //启动压缩
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        ImageStaticHolder.clearImages();
    }
}

File: imagepicker/src/main/java/com/weimu/library/ui/ImageSelectorActivity.kt
package com.weimu.library.ui


import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.constraint.ConstraintLayout
import android.support.v4.content.ContextCompat
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.widget.CheckBox
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import com.weimu.library.ImageStaticHolder
import com.weimu.library.R
import com.weimu.library.adapter.ImageFolderAdapter
import com.weimu.library.adapter.ImageListAdapter
import com.weimu.library.model.LocalMedia
import com.weimu.library.model.LocalMediaFolder
import com.weimu.library.utils.FileUtilsIP
import com.weimu.library.utils.LocalMediaLoader
import com.weimu.universalib.ktx.dip2px
import com.weimu.universalview.core.recyclerview.decoration.GridItemDecoration
import com.weimu.universalview.core.toolbar.StatusBarManager
import com.weimu.universalview.core.toolbar.ToolBarManager
import com.weimu.universalview.ktx.init
import com.weimu.universalview.ktx.setOnClickListenerPro
import com.weimu.universalview.ktx.setTextColorV2
import top.zibin.luban.Luban
import top.zibin.luban.OnCompressListener
import java.io.File
import java.util.*


class ImageSelectorActivity : SelectorBaseActivity() {

    private var maxSelectNum = 9
    private var selectMode = MODE_MULTIPLE
    private var enableCamera = true
    private var enablePreview = true
    private var enableCrop = false
    private var enableCompress = false//是否显示原图按钮

    private val spanCount = 4
    //ui
    private var recyclerView: RecyclerView? = null
    private var imageAdapter: ImageListAdapter? = null
    private var folderLayout: LinearLayout? = null
    private var folderName: TextView? = null
    private var folderWindow: FolderWindow? = null
    private var cbOrigin: CheckBox? = null
    private lateinit var toolBarManager: ToolBarManager

    private var cameraPath: String? = null

    private var allFolders: List<LocalMediaFolder> = arrayListOf()//所有图片文件夹

    private var isUseOrigin = false//是否使用原图


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_imageselector)

        maxSelectNum = intent.getIntExtra(EXTRA_MAX_SELECT_NUM, 9)
        selectMode = intent.getIntExtra(EXTRA_SELECT_MODE, MODE_MULTIPLE)
        enableCamera = intent.getBooleanExtra(EXTRA_SHOW_CAMERA, true)
        enablePreview = intent.getBooleanExtra(EXTRA_ENABLE_PREVIEW, true)
        enableCrop = intent.getBooleanExtra(EXTRA_ENABLE_CROP, false)
        enableCompress = intent.getBooleanExtra(EXTRA_ENABLE_COMPRESS, false)


        if (selectMode == MODE_MULTIPLE) {
            enableCrop = false
        } else {
            enablePreview = false
        }
        if (savedInstanceState != null) {
            cameraPath = savedInstanceState.getString(BUNDLE_CAMERA_PATH)
        }
        initView()
        registerListener()

        ///load data

        LocalMediaLoader(this, LocalMediaLoader.TYPE_IMAGE).loadAllImage(object : LocalMediaLoader.LocalMediaLoadListener {
            override fun loadComplete(folders: List<LocalMediaFolder>) {
                allFolders = folders
                folderWindow!!.bindFolder(allFolders)
                //load all images first
                imageAdapter!!.bindImages(allFolders[0].images as ArrayList<LocalMedia>)
            }
        })
    }

    fun initView() {
        StatusBarManager.setColor(this.window, ContextCompat.getColor(this, R.color.white))
        StatusBarManager.setLightMode(this.window, false)
        toolBarManager = ToolBarManager.with(this, contentView)
                .bg {
                    this.setBackgroundResource(R.color.white)
                }
                .leftMenuIcon {
                    this.setImageResource(R.drawable.toolbar_arrow_back_black)
                }
                .title {
                    this.text = "选择图片"
                }
                .rightMenuText {
                    this.text = if (selectMode == MODE_MULTIPLE) (getString(R.string.done)) else ""
                    this.setTextColorV2(R.color.colorAccent)
                    this.setTextColor(ContextCompat.getColorStateList(context, R.color.black_text_selector))
                    this.isEnabled = false
                    this.setOnClickListenerPro {
                        //点击完成
                        onSelectDone(imageAdapter!!.selectedImages)
                    }
                }

        folderWindow = FolderWindow(this)

        //是否使用原图
        cbOrigin = findViewById(R.id.cb_origin)

        if (!enableCompress) {
            cbOrigin!!.visibility = View.GONE
        }

        cbOrigin!!.isChecked = isUseOrigin
        cbOrigin!!.setOnClickListener { isUseOrigin = !cbOrigin!!.isChecked }

        folderLayout = findViewById<View>(R.id.folder_layout) as LinearLayout
        folderName = findViewById<View>(R.id.folder_name) as TextView

        recyclerView = findViewById<View>(R.id.folder_list) as RecyclerView
        recyclerView?.init()
        recyclerView?.layoutManager = GridLayoutManager(this, spanCount)
        recyclerView?.setHasFixedSize(true)
        recyclerView?.addItemDecoration(GridItemDecoration(spanCount, dip2px(2f), dip2px(2f)))


        imageAdapter = ImageListAdapter(this, maxSelectNum, selectMode, enableCamera, enablePreview)
        recyclerView!!.adapter = imageAdapter
    }

    fun registerListener() {
        folderLayout!!.setOnClickListener(View.OnClickListener {
            //Toast.makeText(ImageSelectorActivity.this, "文件夹长度  " + allFolders.size() + "  内部图片数量  " + allFolders.get(0).getImages().size(), Toast.LENGTH_SHORT).show();
            if (allFolders!!.size == 0 || allFolders!![0].images.size == 0) {
                Toast.makeText(this@ImageSelectorActivity, "没有可选择的图片", Toast.LENGTH_SHORT).show()
                return@OnClickListener
            }
            if (folderWindow!!.isShowing) {
                folderWindow!!.dismiss()
            } else {

                folderWindow!!.showAsDropDown(findViewById<ConstraintLayout>(R.id.cl_toolbar))
            }
        })
        //recyclerView点击事件
        imageAdapter?.setOnImageSelectChangedListener(object : ImageListAdapter.OnImageSelectChangedListener {
            override fun onChange(selectImages: List<LocalMedia>) {
                val enable = selectImages.isNotEmpty()
                if (enable) {
                    toolBarManager.rightMenuText {
                        this.text = "${getString(R.string.done_num)}(${selectImages.size}/${maxSelectNum})"
                        isEnabled = enable
                    }
                } else {
                    toolBarManager.rightMenuText {
                        this.text = getString(R.string.done)
                        isEnabled = enable
                    }
                }
            }

            override fun onTakePhoto() {
                startCamera()
            }

            override fun onPictureClick(media: LocalMedia, position: Int, view: View) {
                if (enablePreview) {

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        startPreviewWithAnim(imageAdapter!!.images, position, view)
                    } else {
                        startPreview(imageAdapter!!.images, position)
                    }

                } else if (enableCrop) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        startCropWithAnim(media.path!!, view)
                    } else {
                        startCrop(media.path)
                    }
                } else {
                    onSelectDone(media.path)
                }
            }
        })
        //点击某个文件件
        folderWindow?.setOnItemClickListener(object : ImageFolderAdapter.OnItemClickListener {
            override fun onItemClick(folderName: String?, images: List<LocalMedia>) {
                folderWindow!!.dismiss()
                imageAdapter!!.bindImages(images as ArrayList<LocalMedia>)
                this@ImageSelectorActivity.folderName!!.text = folderName
                recyclerView!!.smoothScrollToPosition(0)
            }

        })
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            // on take photo success
            if (requestCode == REQUEST_CAMERA) {
                sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(File(cameraPath))))
                if (enableCrop) {
                    startCrop(cameraPath)
                } else {
                    onSelectDone(cameraPath)
                }
            } else if (requestCode == ImagePreviewActivity.REQUEST_PREVIEW) {
                val isDone = data!!.getBooleanExtra(ImagePreviewActivity.OUTPUT_ISDONE, false)
                val images = data.getSerializableExtra(ImagePreviewActivity.OUTPUT_LIST) as List<LocalMedia>
                if (isDone) {
                    onSelectDone(images)
                } else {
                    imageAdapter!!.bindSelectImages(images as ArrayList<LocalMedia>)
                }
            } else if (requestCode == ImageCropActivity.REQUEST_CROP) {
                val path = data!!.getStringExtra(ImageCropActivity.OUTPUT_PATH)
                onSelectDone(path)
            }// on crop success
            //on preview select change
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(BUNDLE_CAMERA_PATH, cameraPath)
    }

    /**
     * start to camera、preview、crop
     */
    fun startCamera() {
        val cameraFile = FileUtilsIP.createCameraFile(this)
        cameraPath = cameraFile.absolutePath
        FileUtilsIP.startActionCapture(this, cameraFile, REQUEST_CAMERA)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun startPreviewWithAnim(previewImages: List<LocalMedia>, position: Int, view: View) {
        ImagePreviewActivity.startPreviewWithAnim(this, imageAdapter!!.selectedImages, maxSelectNum, position, view)
    }

    fun startPreview(previewImages: List<LocalMedia>, position: Int) {
        ImagePreviewActivity.startPreview(this, imageAdapter!!.selectedImages, maxSelectNum, position)
    }

    @SuppressLint("RestrictedApi")
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun startCropWithAnim(path: String, view: View) {
        startActivityForResult(ImageCropActivity.newIntent(this, path), ImageCropActivity.REQUEST_CROP,
                ActivityOptions.makeSceneTransitionAnimation(this, view, "share_image").toBundle())
    }

    fun startCrop(path: String?) {
        startActivityForResult(ImageCropActivity.newIntent(this, path!!), ImageCropActivity.REQUEST_CROP)
    }

    /**
     * on select done
     *
     * @param medias
     */
    fun onSelectDone(medias: List<LocalMedia>) {
        val images = ArrayList<String>()
        for (media in medias) {
            images.add(media.path!!)
        }
        onResult(images)
    }

    fun onSelectDone(path: String?) {
        val images = ArrayList<String>()
        images.add(path!!)
        onResult(images)
    }

    //返回图片
    fun onResult(images: ArrayList<String>) {
        if (isUseOrigin) {
            setResult(Activity.RESULT_OK, Intent().putStringArrayListExtra(REQUEST_OUTPUT, images))
            finish()
        } else {
            compressImage(images)

        }
    }

    //压缩图片
    private fun compressImage(photos: ArrayList<String>) {
        //Toast.makeText(this, "压缩中...", Toast.LENGTH_SHORT).show();
        val newImageList = ArrayList<String>()
        Luban.with(this)
                .load(photos)                                   // 传人要压缩的图片列表
                .ignoreBy(100)                                  // 忽略不压缩图片的大小
                .setCompressListener(object : OnCompressListener { //设置回调
                    override fun onStart() {
                        Log.d("weimu", "开始压缩")
                    }

                    override fun onSuccess(file: File) {
                        Log.d("weimu", "压缩成功 地址为：$file")
                        newImageList.add(file.toString())
                        //所有图片压缩成功
                        if (newImageList.size == photos.size) {
                            setResult(Activity.RESULT_OK, Intent().putStringArrayListExtra(REQUEST_OUTPUT, newImageList))
                            finish()
                        }
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                    }
                }).launch()    //启动压缩
    }


    override fun onDestroy() {
        super.onDestroy()
        ImageStaticHolder.clearImages()
    }

    companion object {
        val REQUEST_IMAGE = 66
        val REQUEST_CAMERA = 67

        val BUNDLE_CAMERA_PATH = "CameraPath"

        val REQUEST_OUTPUT = "outputList"


        val EXTRA_MAX_SELECT_NUM = "MaxSelectNum"//最大选择数
        val EXTRA_SELECT_MODE = "SelectMode"//选择模式
        val EXTRA_SHOW_CAMERA = "ShowCamera"//是否显示摄像头
        val EXTRA_ENABLE_PREVIEW = "EnablePreview"//是否需要预览
        val EXTRA_ENABLE_CROP = "EnableCrop"//是否需要裁剪
        val EXTRA_ENABLE_COMPRESS = "EnableCompress"//是否需要压缩

        val MODE_MULTIPLE = 1
        val MODE_SINGLE = 2

        fun start(activity: Activity, maxSelectNum: Int, mode: Int, enableCamera: Boolean, enablePreview: Boolean, enableCrop: Boolean, enableCompress: Boolean) {
            val intent = Intent(activity, ImageSelectorActivity::class.java)
            intent.putExtra(EXTRA_MAX_SELECT_NUM, maxSelectNum)
            intent.putExtra(EXTRA_SELECT_MODE, mode)
            intent.putExtra(EXTRA_SHOW_CAMERA, enableCamera)
            intent.putExtra(EXTRA_ENABLE_PREVIEW, enablePreview)
            intent.putExtra(EXTRA_ENABLE_CROP, enableCrop)
            intent.putExtra(EXTRA_ENABLE_COMPRESS, enableCompress)
            activity.startActivityForResult(intent, REQUEST_IMAGE)
        }
    }
}
